server/mqtt.go

```diff
@@ -2591,65 +2697,86 @@ func (as *mqttAccountSessionManager) serializeRetainedMsgsForSub(rms map[string]
 		sub.mqtt.prm = append(sub.mqtt.prm, headerBytes, rm.Msg)
 		c.mu.Unlock()
 		if trace {
-			pp := mqttPublish{
+			toTrace = append(toTrace, mqttPublish{
 				topic: []byte(rm.Topic),
 				flags: flags,
 				pi:    pi,
 				sz:    len(rm.Msg),
-			}
-			c.traceOutOp("PUBLISH", []byte(mqttPubTrace(&pp)))
+			})
 		}
 	}
+	for _, pp := range toTrace {
+		c.traceOutOp("PUBLISH", []byte(mqttPubTrace(&pp)))
+	}
+	return nil
 }
 
-// Returns in the provided slice all publish retained message records that
+// Appends the stored message subjects for all retained message records that
 // match the given subscription's `subject` (which could have wildcards).
 //
 // Account session manager NOT lock held on entry.
-func (as *mqttAccountSessionManager) loadRetainedMessagesForSubject(rms map[string]*mqttRetainedMsg, topSubject []byte, log Logger) {
+func (as *mqttAccountSessionManager) addRetainedSubjectsForSubject(list map[string]struct{}, topSubject string) bool {
 	as.mu.RLock()
 	if len(as.retmsgs) == 0 {
 		as.mu.RUnlock()
-		return
+		return false
 	}
 	result := as.sl.ReverseMatch(string(topSubject))
 	as.mu.RUnlock()
 
-	if len(result.psubs) == 0 {
-		return
-	}
+	added := false
 	for _, sub := range result.psubs {
 		subject := string(sub.subject)
-		if rms[subject] != nil {
-			continue // already loaded
+		if _, ok := list[subject]; ok {
+			continue
 		}
+		list[subject] = struct{}{}
+		added = true
+	}
 
-		// See if we have the retained message in the cache.
-		if rmv, _ := as.rmsCache.Load(subject); rmv != nil {
-			rm := rmv.(mqttRetainedMsg)
-			rms[subject] = &rm
-			continue
+	return added
+}
+
+type warner interface {
+	Warnf(format string, v ...any)
+}
+
+// Loads a list of retained messages given a list of stored message subjects.
+func (as *mqttAccountSessionManager) loadRetainedMessages(subjects map[string]struct{}, w warner) map[string]*mqttRetainedMsg {
+	rms := make(map[string]*mqttRetainedMsg)
```

Comment:
It is assumed that `rms` length will be of `subjects` length. So you could initialize with that, same for `ss`.